File: solutions/4285-smallest-stable-index-ii/smallest-stable-index-ii.py
class Solution:
    def firstStableIndex(self, nums, k):
        n = len(nums)
        prefixMax = [0] * n
        suffixMin = [10**9] * n
        prefixMax[0] = nums[0]
        suffixMin[n - 1] = nums[n - 1]
        for i in range(1, n):
            prefixMax[i] = max(prefixMax[i - 1], nums[i])
            suffixMin[n - 1 - i] = min(suffixMin[n - i], nums[n - 1 - i])
        for i in range(n):
            if prefixMax[i] - suffixMin[i] <= k:
                return i
        return -1

File: solutions/4285-smallest-stable-index-ii/smallest-stable-index-ii.go
func firstStableIndex(nums []int, k int) int {
	n := len(nums)
	minValue := make([]int, n)
	minValue[n-1] = nums[n-1]
	for i := n - 2; i >= 0; i-- {
		if minValue[i+1] < nums[i] {
			minValue[i] = minValue[i+1]
		} else {
			minValue[i] = nums[i]
		}
	}

	maxValue := 0
	for i := 0; i < n; i++ {
		if nums[i] > maxValue {
			maxValue = nums[i]
		}
		if maxValue-minValue[i] <= k {
			return i
		}
	}
	return -1
}